pkg/cmd/release-payload-controller/payload_creation_controller.go

```diff
@@ -0,0 +1,230 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"github.com/openshift/library-go/pkg/operator/events"
+)
+
+// These are valid Reason values for the ReleasePayloadStatus Conditions
+const (
+	// ReleasePayloadCreatedReason programmatic identifier indicating that the ReleasePayload created successfully
+	ReleasePayloadCreatedReason string = "ReleasePayloadCreated"
+
+	// ReleasePayloadFailedReason programmatic identifier indicating that the ReleasePayload creation failed
+	ReleasePayloadFailedReason string = "ReleasePayloadFailed"
+)
+
+// PayloadCreationController is responsible for setting the PayloadCreated and PayloadFailed conditions based
+// on whether the release payload creation job completed successfully or not.
+// The PayloadCreationController reads the following pieces of information:
+//   - .status.releaseCreationJobResult.status
+// and populates the following conditions:
+//   - .status.conditions.PayloadCreated
+//   - .status.conditions.PayloadFailed
+type PayloadCreationController struct {
+	releasePayloadNamespace string
+	releasePayloadLister    releasepayloadlister.ReleasePayloadLister
+	releasePayloadClient    releasepayloadclient.ReleaseV1alpha1Interface
+
+	eventRecorder events.Recorder
+
+	cachesToSync []cache.InformerSynced
+
+	queue workqueue.RateLimitingInterface
+}
+
+func NewPayloadCreationController(
+	releasePayloadNamespace string,
+	releasePayloadInformer releasepayloadinformer.ReleasePayloadInformer,
+	releasePayloadClient releasepayloadclient.ReleaseV1alpha1Interface,
+	eventRecorder events.Recorder,
+) (*PayloadCreationController, error) {
+	c := &PayloadCreationController{
+		releasePayloadNamespace: releasePayloadNamespace,
+		releasePayloadLister:    releasePayloadInformer.Lister(),
+		releasePayloadClient:    releasePayloadClient,
+		eventRecorder:           eventRecorder.WithComponentSuffix("payload-creation-controller"),
+		queue:                   workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "PayloadCreationController"),
+	}
+
+	c.cachesToSync = append(c.cachesToSync, releasePayloadInformer.Informer().HasSynced)
+
+	releasePayloadFilter := func(obj interface{}) bool {
+		if releasePayload, ok := obj.(*v1alpha1.ReleasePayload); ok {
+			// If the conditions are both in their respective terminal states, then there is nothing else to do...
+			if (conditions.IsConditionTrue(releasePayload.Status.Conditions, v1alpha1.ConditionPayloadCreated) || conditions.IsConditionFalse(releasePayload.Status.Conditions, v1alpha1.ConditionPayloadCreated)) &&
+				(conditions.IsConditionTrue(releasePayload.Status.Conditions, v1alpha1.ConditionPayloadFailed) || conditions.IsConditionFalse(releasePayload.Status.Conditions, v1alpha1.ConditionPayloadFailed)) {
+				return false
+			}
+			return true
+		}
+		if tombstone, ok := obj.(cache.DeletedFinalStateUnknown); ok {
```

Comment:
unnecessary block.  Can be removed in followup

Reply:
Added a TODO

Reply:
Removed
Done